# Probability

## Introduction to Probability

Probability is a measure of how likely an event is to happen. Its value ranges from 0 to 1, where 0 means impossible and 1 means certain.

### Formula
P(A)=Number of Favorable Outcomes/Total Number of Possible Outcomes

### Numerical Example

A fair coin is tossed once.

Possible outcomes = Head, Tail

Favorable outcomes for Head = 1
Total outcomes = 2

Therefore:

P(Head)= 1/2

=0.5

### AI/ML Engineer Example

Probability is used extensively in ML to predict the likelihood of outcomes, such as whether an email is spam or whether a customer will purchase a product.

In [2]:
#Python Implementation

favorable_outcomes = 1
total_outcomes = 6

probability = favorable_outcomes / total_outcomes

print("Probability:", probability)

Probability: 0.16666666666666666


In the Python code, we store the number of favorable outcomes as 1 and the total number of outcomes as 6. We divide the favorable outcomes by the total outcomes to calculate the probability. The result is approximately 0.167, or 16.67%.

## Events

An event is a specific outcome or a group of outcomes that we are interested in during a probability experiment. An event can contain one or more possible outcomes.

### Formula
P(A)= Favorable Outcomes/Total Outcomes


### Numerical Example

When rolling a dice, the event of getting an even number is:

 A=\{2,4,6\}

Therefore:

P(A)= 3/6

=0.5

### AI/ML Engineer Example

In ML, an event can represent an outcome such as a customer purchasing a product or an email being classified as spam.

In [3]:
### Python Implementation
outcomes = [1, 2, 3, 4, 5, 6]

even_numbers = [x for x in outcomes if x % 2 == 0]

probability = len(even_numbers) / len(outcomes)

print("Even Numbers:", even_numbers)
print("Probability:", probability)

Even Numbers: [2, 4, 6]
Probability: 0.5


In the Python code, we first create a list containing all six dice outcomes. The list comprehension checks each number using x % 2 == 0 and keeps only the even numbers. We then divide the number of favorable outcomes, 3, by the total outcomes, 6, giving a probability of 0.5 or 50%.

## Sample Space

A sample space is the complete set of all possible outcomes of a probability experiment. It is usually represented by the symbol S.

### There is no specific calculation formula.

S={All Possible Outcomes}

### Formula

There is no specific calculation formula.

S=\{\text{All Possible Outcomes}\} 

###AI/ML Engineer Example

An ML engineer can define a sample space for a classification problem, such as {Spam, Not Spam} for email classification.

In [4]:
#Python Implementation
sample_space = ["Head", "Tail"]

print("Sample Space:", sample_space)

Sample Space: ['Head', 'Tail']


In the Python code, we create a list called sample_space and store "Head" and "Tail" in it. These represent all possible outcomes of one coin toss, so the list represents the complete sample space.

## Independent Events

Two events are independent when the occurrence of one event does not affect the probability of the other event. Tossing a coin twice is an example because the first toss does not affect the second toss.

### Formula

Formula P(A and B) = P(A) × P(B)

### Numerical Example
Suppose we toss a coin and roll a die.

Event A = getting Head.

P(A) = 1/2

Event B = getting 6.

P(B) = 1/6

The coin result does not affect the die result.

Therefore:

P(A and B) = 1/2 × 1/6

= 1/12

### AI/ML Engineer Example

Independent events are useful when an ML model assumes that certain observations or features are independent of each other, such as in some probabilistic models.

In [6]:
#Python Implementation

p_head_1 = 0.5
p_head_2 = 0.5

probability = p_head_1 * p_head_2

print("Probability:", probability)

Probability: 0.25


In the Python code, we store the probability of getting Head on each coin as 0.5. Since the events are independent, we multiply the two probabilities. The result is 0.25, meaning there is a 25% probability of getting Head on both coins.

## Dependent Events

Two events are dependent when the occurrence of one event changes the probability of the other event. This commonly happens when objects are selected without replacement

### Formula
P(A∩B)=P(A)×P(B∣A)

### Numerical Example

Numerical Example
A bag contains:

3 red balls 2 blue balls

Total = 5 balls.

Event A = selecting a red ball first.

P(A) = 3/5

After selecting one red ball, the bag contains:

2 red balls 2 blue balls

Now:

P(B|A) = 2/4

Therefore:

P(A and B) = 3/5 × 2/4

= 6/20

= 0.30

### AI/ML Engineer Example

Dependent events are important when previous observations affect future observations, such as sequential customer behavior or sampling without replacement.

In [7]:
#Python Implementation

red_balls = 3
total_balls = 5

p_first_red = red_balls / total_balls

red_balls_after = 2
total_balls_after = 4

p_second_red = red_balls_after / total_balls_after

probability = p_first_red * p_second_red

print("Probability:", probability)

Probability: 0.3


In the Python code, we first calculate the probability of selecting a red ball as 3/5. After selecting one red ball, only 2 red balls remain out of 4 total balls, so the second probability becomes 2/4. We multiply both probabilities to get 0.3 or 30%.

## Conditional Probability

Conditional probability measures the probability of one event occurring when we already know that another event has occurred. It allows us to update a probability using additional information.

### Formula
P(A|B) = P(A ∩ B) / P(B)

Where:

P(A|B) = Probability of A given B

P(A ∩ B) = Probability of both A and B

P(B) = Probability of B

### Numerical Example

Suppose:

P(A ∩ B) = 0.3
P(B) = 0.6

Then:

P(A∣B)= 0.3/0.6

=0.5

### AI/ML Engineer Example

Conditional probability is widely used in ML to calculate probabilities such as the probability that an email is spam given that certain words appear in it.

In [8]:
#Python Implementation
p_a_and_b = 0.3
p_b = 0.6

conditional_probability = p_a_and_b / p_b

print("Conditional Probability:", conditional_probability)

Conditional Probability: 0.5


In the Python code, p_a_and_b represents the probability that both events occur, while p_b represents the probability that the customer received the email. We divide the joint probability by the probability of the known event to calculate the conditional probability, giving 0.5 or 50%.

## Bayes' Theorem
Definition

Bayes' Theorem calculates the probability of an event based on new evidence or information. It helps us update an existing probability when new evidence becomes available.

### Formula 
P(A|B) = [P(B|A) × P(A)] / P(B)

Where:

P(A|B) = Posterior probability

P(B|A) = Likelihood

P(A) = Prior probability

P(B) = Probability of evidence

### AI/ML Engineer Example

Bayes' theorem is the foundation of Naive Bayes classification, where an ML engineer calculates the probability of a class based on observed features.

In [9]:
#Python Implementation
p_a = 0.2
p_b_given_a = 0.8
p_b = 0.4

p_a_given_b = (p_b_given_a * p_a) / p_b

print("P(A|B):", p_a_given_b)

P(A|B): 0.4000000000000001


In the Python code, we store the prior probability p_a, the probability of observing the evidence when A is true p_b_given_a, and the overall probability of the evidence p_b. We then apply Bayes' formula to calculate the updated probability, which is 0.4 or 40%.

## Probability Rules

Probability rules are mathematical rules used to calculate probabilities for different combinations of events. Common rules include the complement, addition, and multiplication rules.

Rule 1 – Probability Range 0 ≤ P(A) ≤ 1

Rule 2 – Complement Rule P(Aᶜ) = 1 - P(A)

Rule 3 – Addition Rule For mutually exclusive events:

P(A ∪ B) = P(A) + P(B)

For general events:

P(A ∪ B) = P(A) + P(B) - P(A ∩ B)

Rule 4 – Multiplication Rule For independent events:

P(A ∩ B) = P(A) × P(B)

For dependent events:

P(A ∩ B) = P(A) × P(B|A)

### Numerical Example 
Suppose:

P(A) = 0.6

Using the complement rule:

P(Aᶜ) = 1 - 0.6

P(Aᶜ) = 0.4

Therefore, if the probability of a customer purchasing a product is 60%, the probability of the customer not purchasing it is 40%.

Real-World Example A weather service may report a 70% probability of rain.

The probability of no rain is:

1 - 0.70 = 0.30

### AI/ML Engineer Example

Probability rules are used throughout ML for calculating joint probabilities, conditional probabilities, and probabilities of different prediction outcomes.

In [10]:
# Python Implementation
p_a = 0.6

p_not_a = 1 - p_a

print("Probability of not A:", p_not_a)

Probability of not A: 0.4


In the Python code, we store the probability of purchase as 0.6. Since the probability of an event and its complement must add up to 1, we subtract 0.6 from 1. The result is 0.4 or 40%, which is the probability that the customer will not purchase.

## Random Variables

A random variable is a variable whose value depends on the outcome of a random experiment. It can be discrete, such as the number of customers, or continuous, such as temperature or height.

### Formula

For a discrete random variable:

P(X=x)

### AI/ML Engineer Example

ML engineers use random variables to represent uncertain quantities such as customer demand, predicted classes, stock prices, or model errors.

In [11]:
#Python Implementation
import random

X = random.randint(1, 6)

print("Random Variable X:", X)

Random Variable X: 6


In the Python code, random.randint(1, 6) randomly generates an integer between 1 and 6. The generated number is stored in the variable X. Since the result is random, X can have a different value each time the program runs.

## Expected Value

Expected Value is the long-term average value of a random variable when an experiment is repeated many times. It is calculated by multiplying each possible outcome by its probability and adding the results.

### Formula 
E(X) = Σ [x × P(X = x)]

Where:

x = Possible outcome

P(X = x) = Probability of that outcome

### Numerical Example 
Suppose a game has two possible outcomes:

Win = ₹100 with probability 0.4

Lose = ₹0 with probability 0.6

Expected Value:

E(X) = (100 × 0.4) + (0 × 0.6)

E(X) = 40

The expected value is ₹40.

This does not mean that every game will give exactly ₹40.

It means that over many repeated games, the average outcome approaches ₹40 under the stated probabilities.

### AI/ML Engineer Example

An ML engineer can use expected value to calculate the expected outcome or expected loss of a model, such as expected customer revenue or the average cost of incorrect predictions.

In [12]:
#Python Implementation
outcomes = [100, 200, 500]
probabilities = [0.5, 0.3, 0.2]

expected_value = sum(
    outcome * probability
    for outcome, probability in zip(outcomes, probabilities)
)

print("Expected Value:", expected_value)

Expected Value: 210.0


In the Python code, we store the possible revenue values in outcomes and their probabilities in probabilities. The zip() function pairs each revenue with its corresponding probability. We multiply each outcome by its probability and use sum() to add all the weighted values. The expected value is ₹210, representing the long-term average outcome.